### PR TITLE
file: 5.44 -> 5.45

### PR DESCRIPTION
**Pull request by trofi**
Changes: https://github.com/file/file/compare/FILE5_44...FILE5_45

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
Fails some tests. Needs a bit more investigation.

**Comment by trofi** [comment]
`pkgsi686Linux.file` fails some date comparison tests.

**Comment by trofi** [comment]
Fixed i686 tests. Ready for review.

File: pkgs/tools/misc/file/default.nix [hunk 1]
```diff
@@ -7,31 +7,24 @@
...
-    ./pyzip.patch
-
-    # Backport fix for --uncompress always detecting contents as "empty"
-    (fetchurl {
-      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-apps/file/files/file-5.44-decompress-empty.patch?h=dfc57da515a2aaf085bea68267cc727f1bfaa691";
-      hash = "sha256-fUzRQAlLWczBmR5iA1Gk66mHjP40MJcMdgCtm2+u1SQ=";
-    })
+    # Upstream patch to fix 32-bit tests.
+    # WIll be included in 5.46+ releases.
+    ./32-bit-time_t.patch
```
Comment (doronbehar) [inline pkgs/tools/misc/file/default.nix:23]: Why not `fetchpatch`?

File: pkgs/tools/misc/file/default.nix [hunk 2]
```diff
@@ -7,31 +7,24 @@
...
   ];
 
+  outputs = [ "out" "dev" "man" ];
```
Comment (doronbehar) [inline pkgs/tools/misc/file/default.nix:26]: Nice!

**Review by doronbehar — CHANGES_REQUESTED** [review]

File: pkgs/tools/misc/file/default.nix (hunk 2 above)
Comment (trofi) [reply pkgs/tools/misc/file/default.nix]: Not a new change I'm afraid. It's a bit confusing diff due to me plugging `patches` in a slightly different place. Reordered to make it more obvious.

File: pkgs/tools/misc/file/default.nix (hunk 1 above)
Comment (trofi) [reply pkgs/tools/misc/file/default.nix:25]: `fetchpatch` has `file` in it's dependencies. As a result we can't use it for most of packages in `stdenv` closure:

```diff
--- a/pkgs/tools/misc/file/default.nix
+++ b/pkgs/tools/misc/file/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchurl, file, zlib, libgnurx }:
+{ lib, stdenv, fetchurl, fetchpatch, file, zlib, libgnurx }:

 # Note: this package is used for bootstrapping fetchurl, and thus
 # cannot use fetchpatch! All mutable patches (generated by GitHub or
@@ -22,7 +22,11 @@ stdenv.mkDerivation rec {
   patches = [
     # Upstream patch to fix 32-bit tests.
     # WIll be included in 5.46+ releases.
-    ./32-bit-time_t.patch
+    # ./32-bit-time_t.patch
+    (fetchpatch {
+      url = "https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch";
+      hash = "sha256-/Jf1ECm1DiyfTjv/79r2ePDgOe6HK53lwAKm0Jx4TYI=";
+    })
   ];

   strictDeps = true;
```

```
$ nix build -f. file
error:
       … while evaluating a branch condition

         at /mnt/archive/big/git/nixpkgs-master/pkgs/stdenv/booter.nix:99:7:

           98|     thisStage =
           99|       if args.__raw or false
             |       ^
          100|       then args'

       … in the right operand of the update (//) operator

         at /mnt/archive/big/git/nixpkgs-master/pkgs/stdenv/booter.nix:84:7:

           83|       { allowCustomOverrides = index == 1; }
           84|       // (stageFun prevStage))
             |       ^
           85|     (lib.lists.reverseList stageFuns);

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: function 'anonymous lambda' called with unexpected argument 'postFetch'

       at /mnt/archive/big/git/nixpkgs-master/pkgs/build-support/fetchurl/boot.nix:5:1:

            4|
            5| { url ? builtins.head urls
             | ^
            6| , urls ? []
```

Comment (doronbehar) [reply pkgs/tools/misc/file/default.nix:25]: And what about `fetchurl`? That was used before to fetch the gentoo patch.

Comment (trofi) [reply pkgs/tools/misc/file/default.nix:25]: `fetchurl` would be fine if we fetched a file with known contents upfront. The gentoo patch was a static file in their repository.

Unfortunately links like https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch are not guaranteed to be stable from one to another `github` release and thus `fetchurl` will fail it's hash from time to time (when metadata around the patch changes). `fetchpatch` is usually a solution for cases like those as it strips metadata first and then calculates the hash). But `fetchpatch` is not available for `stdenv` closure.

Comment (amjoseph-nixpkgs) [reply pkgs/tools/misc/file/default.nix:25]: @trofi is correct.  This is an unfortunate limitation of nixpkgs (and nix, [except very bleeding-edge versions](https://github.com/NixOS/nixpkgs/pull/188587)) right now.

File: pkgs/tools/misc/file/default.nix [hunk 3]
```diff
@@ -7,29 +7,22 @@
...
-    # https://salsa.debian.org/reproducible-builds/strip-nondeterminism/-/issues/20
-    ./pyzip.patch
-
-    # Backport fix for --uncompress always detecting contents as "empty"
-    (fetchurl {
-      url = "https://gitweb.gentoo.org/repo/gentoo.git/plain/sys-apps/file/files/file-5.44-decompress-empty.patch?h=dfc57da515a2aaf085bea68267cc727f1bfaa691";
-      hash = "sha256-fUzRQAlLWczBmR5iA1Gk66mHjP40MJcMdgCtm2+u1SQ=";
-    })
+    # Upstream patch to fix 32-bit tests.
+    # WIll be included in 5.46+ releases.
```
Comment (amjoseph-nixpkgs) [inline pkgs/tools/misc/file/default.nix:24]:
```suggestion
    # Will be included in 5.46+ releases.
```

File: pkgs/tools/misc/file/default.nix (hunk 1 above)
Comment (doronbehar) [reply pkgs/tools/misc/file/default.nix:25]: > Unfortunately links like [file/file@`218fdf8`.patch](https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.patch) are not guaranteed to be stable from one to another `github` release

Interesting.. Any idea why is that? What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

Comment (trofi) [reply pkgs/tools/misc/file/default.nix:25]: > Interesting.. Any idea why is that?

Good question. I would guess that long-tern stability is not their priority (just like `git diff` or `patch` slightly change their output from release to release).

> What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

Yeah, those look better. It still has a chance break in future around `index 2e0494d2f..78f574ea1 100644` and use more than 9 digits of abbreviation. I guess digit count is dynamically increasing with repository's commit count.

For example `linux` uses 14-digit abbreviation right now: https://github.com/torvalds/linux/commit/024ff300db33968c133435a146d51ac22db27374.diff
`index 1fbda2f94184e3..b21335e6a210e6 100644`

And `gcc` uses 12 digits: https://github.com/gcc-mirror/gcc/commit/5b42ee2cded7c2173ea0e9f9e08dd1d9b226ed3e.diff `index 181df09e5879..67ac03558b6a 100644`

Comment (amjoseph-nixpkgs) [reply pkgs/tools/misc/file/default.nix:25]: > Interesting.. Any idea why is that? What about https://github.com/file/file/commit/218fdf813fd5ccecbb8887a1b62509cd1c6dd3a1.diff ? Seems without metadata to me.

@doronbehar, see the comment at the top of `fetchpatch/default.nix`.

**Review by doronbehar — APPROVED** [review]

**Comment by amjoseph-nixpkgs** [comment]
@ofborg eval

**Comment by trofi** [comment]
`python3Packages.python-magic` testsuite regressed. Proposed https://github.com/NixOS/nixpkgs/pull/250362 to fix it.

**Pull request merged**